File: frontend/src/types/pool-allocation.ts
/**
 * PoolAllocation domain types for split expense allocations.
 *
 * These Zod schemas match exactly with backend/app/models/pool_allocation.py.
 * Pool allocations handle splitting expenses from a source pool
 * to multiple target pools with percentage or fixed amount allocations.
 */

import { z } from 'zod'

/**
 * Helper for allocation value validation.
 * For percentage: must be between 0 and 100
 * For fixed_amount: must be positive
 */
const allocationValueString = z
  .string()
  .regex(/^\d+(\.\d+)?$/, 'Must be a valid decimal number')
  .refine(
    (val) => {
      const num = parseFloat(val)
      return num > 0
    },
    { message: 'Allocation value must be positive' }
  )

/**
 * AllocationType enum matching backend.
 */
const AllocationTypeSchema = z.enum(['percentage', 'fixed_amount'])

/**
 * Full PoolAllocation model from database.
 *
 * Represents a single allocation from a source pool to a target pool.
 */
export const PoolAllocationSchema = z.object({
  id: z.string().uuid(),
  source_pool_id: z.string().uuid(),
  target_pool_id: z.string().uuid(),
  allocation_type: AllocationTypeSchema,
  allocation_value: allocationValueString,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
})

export type PoolAllocation = z.infer<typeof PoolAllocationSchema>

export interface PoolAllocationListResponse {
  data: PoolAllocation[]
  count: number
  has_more: boolean
}

/**
 * DTO for creating a pool allocation.
 */
export const PoolAllocationCreateSchema = z
  .object({
    source_pool_id: z.string().uuid(),
    target_pool_id: z.string().uuid(),
    allocation_type: AllocationTypeSchema,
    allocation_value: allocationValueString,
  })
  .refine(
    (data) => {
      // Validate percentage range
      if (data.allocation_type === 'percentage') {
        const value = parseFloat(data.allocation_value)
        return value > 0 && value <= 100
      }
      // Fixed amounts just need to be positive (already validated by allocationValueString)
      return true
    },
    {
      message: 'Percentage allocation must be between 0 and 100',
      path: ['allocation_value'],
    }
  )

export type PoolAllocationCreate = z.infer<typeof PoolAllocationCreateSchema>

/**
 * DTO for updating a pool allocation.
 *
 * All fields are optional for partial updates.
 */
export const PoolAllocationUpdateSchema = z.object({
  target_pool_id: z.string().uuid().optional(),
  allocation_type: AllocationTypeSchema.optional(),
  allocation_value: allocationValueString.optional(),
})

export type PoolAllocationUpdate = z.infer<typeof PoolAllocationUpdateSchema>

/**
 * Validate that percentage allocations sum to exactly 100%.
 *
 * @param allocations - List of allocation DTOs to validate
 * @returns Object with isValid boolean and error message if invalid
 */
export function validateAllocationsSumTo100(
  allocations: PoolAllocationCreate[]
): { isValid: boolean; error: string } {
  // Filter to only percentage allocations
  const percentageAllocations = allocations.filter(
    (a) => a.allocation_type === 'percentage'
  )

  if (percentageAllocations.length === 0) {
    // No percentage allocations to validate
    return { isValid: true, error: '' }
  }

  // FIX F-106: sum the percentages using integer arithmetic so floating-point
  // accumulation drift can never push a legitimately-100% split outside the
  // tolerance below. We scale every value to integers at the largest decimal
  // precision present, sum those exactly, then divide once at the end (an
  // exact operation for these magnitudes). A naive `sum + parseFloat(...)`
  // reduce accumulates IEEE-754 error across many addends.
  const decimalPlaces = percentageAllocations.reduce((max, a) => {
    const fraction = a.allocation_value.split('.')[1] ?? ''
    return Math.max(max, fraction.length)
  }, 0)
  const factor = 10 ** decimalPlaces
  const totalScaled = percentageAllocations.reduce(
    (sum, a) => sum + Math.round(parseFloat(a.allocation_value) * factor),
    0
  )
  const total = totalScaled / factor

  // Check if sum equals 100 (with tolerance for decimal precision)
  if (Math.abs(total - 100) > 0.01) {
    return {
      isValid: false,
      error: `Percentage allocations must sum to 100%, got ${total.toFixed(2)}%`,
    }
  }

  return { isValid: true, error: '' }
}

File: backend/app/models/pool_allocation.py
"""PoolAllocation domain types for split expense allocations.

These models handle the splitting of expenses from a source pool
to multiple target pools with percentage or fixed amount allocations.
"""

from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field, ValidationInfo, field_validator

from app.models.enums import AllocationType


class PoolAllocation(BaseModel):
    """Split allocation linking an expense pool to target pools.

    Represents a single allocation from a source pool to a target pool.
    Multiple allocations from the same source pool create a split.
    """

    id: UUID
    source_pool_id: UUID = Field(
        ..., description="Pool being split into multiple targets"
    )
    target_pool_id: UUID = Field(
        ..., description="Destination pool receiving allocation"
    )
    allocation_type: AllocationType = Field(
        ..., description="Whether allocation is percentage or fixed amount"
    )
    allocation_value: Decimal = Field(
        ...,
        description=(
            "Percentage (0-100) if percentage type, " "or dollar amount if fixed_amount"
        ),
    )
    created_at: datetime
    updated_at: datetime

    @field_validator("allocation_value")
    @classmethod
    def validate_allocation_value(cls, v: Decimal, info: ValidationInfo) -> Decimal:
        """Validate allocation value based on type.

        For percentage allocations, value must be between 0 and 100.
        For fixed amounts, value must be positive.
        """
        allocation_type = info.data.get("allocation_type")
        if allocation_type == AllocationType.PERCENTAGE:
            if v <= 0 or v > 100:
                raise ValueError("Percentage allocation must be between 0 and 100")
        elif allocation_type == AllocationType.FIXED_AMOUNT:
            if v <= 0:
                raise ValueError("Fixed amount allocation must be positive")
        return v


class PoolAllocationCreate(BaseModel):
    """DTO for creating a pool allocation."""

    source_pool_id: UUID = Field(
        ..., description="Pool being split into multiple targets"
    )
    target_pool_id: UUID = Field(
        ..., description="Destination pool receiving allocation"
    )
    allocation_type: AllocationType = Field(
        ..., description="Whether allocation is percentage or fixed amount"
    )
    allocation_value: Decimal = Field(
        ...,
        description=(
            "Percentage (0-100) if percentage type, " "or dollar amount if fixed_amount"
        ),
    )

    @field_validator("allocation_value")
    @classmethod
    def validate_allocation_value(cls, v: Decimal, info: ValidationInfo) -> Decimal:
        """Validate allocation value based on type.

        For percentage allocations, value must be between 0 and 100.
        For fixed amounts, value must be positive.
        """
        allocation_type = info.data.get("allocation_type")
        if allocation_type == AllocationType.PERCENTAGE:
            if v <= 0 or v > 100:
                raise ValueError("Percentage allocation must be between 0 and 100")
        elif allocation_type == AllocationType.FIXED_AMOUNT:
            if v <= 0:
                raise ValueError("Fixed amount allocation must be positive")
        return v


class PoolAllocationUpdate(BaseModel):
    """DTO for updating a pool allocation.

    All fields are optional for partial updates.
    """

    target_pool_id: UUID | None = Field(
        None, description="Destination pool receiving allocation"
    )
    allocation_type: AllocationType | None = Field(
        None, description="Whether allocation is percentage or fixed amount"
    )
    allocation_value: Decimal | None = Field(
        None,
        description=(
            "Percentage (0-100) if percentage type, " "or dollar amount if fixed_amount"
        ),
    )

    @field_validator("allocation_value")
    @classmethod
    def validate_allocation_value(
        cls, v: Decimal | None, info: ValidationInfo
    ) -> Decimal | None:
        """Validate allocation value based on type.

        For percentage allocations, value must be between 0 and 100.
        For fixed amounts, value must be positive.
        """
        if v is None:
            return v

        allocation_type = info.data.get("allocation_type")
        if allocation_type == AllocationType.PERCENTAGE:
            if v <= 0 or v > 100:
                raise ValueError("Percentage allocation must be between 0 and 100")
        elif allocation_type == AllocationType.FIXED_AMOUNT:
            if v <= 0:
                raise ValueError("Fixed amount allocation must be positive")
        return v


def validate_allocations_sum_to_100(
    allocations: list[PoolAllocationCreate],
) -> tuple[bool, str]:
    """Validate that percentage allocations sum to exactly 100%.

    Args:
        allocations: List of allocation DTOs to validate

    Returns:
        Tuple of (is_valid, error_message)
        - is_valid: True if allocations are valid, False otherwise
        - error_message: Empty string if valid, error description if invalid
    """
    # Filter to only percentage allocations
    percentage_allocations = [
        a for a in allocations if a.allocation_type == AllocationType.PERCENTAGE
    ]

    if not percentage_allocations:
        # No percentage allocations to validate
        return True, ""

    # Sum all percentage allocations
    total = sum(a.allocation_value for a in percentage_allocations)

    # Check if sum equals 100 (with tolerance for decimal precision)
    if abs(total - Decimal("100")) > Decimal("0.01"):
        return False, f"Percentage allocations must sum to 100%, got {total}%"

    return True, ""
